src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -396,111 +386,55 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+            ImmutableArray<(bool solutionChanged, ProjectId changedProject)> workQueue, CancellationToken cancellationToken)
         {
-            lock (_gate)
-            {
-                // Augment the data that the foreground thread will process.
-                _solutionChanged |= localSolutionChanged;
-                if (localChangedProject != null)
-                {
-                    _changedProjects.Add(localChangedProject);
-                }
-
-                // Now cancel any inflight work that is processing the data.
-                _tokenSource.Cancel();
-                _tokenSource = new CancellationTokenSource();
-
-                // And enqueue a new job to process things.  Wait one second before starting.
-                // That way if we get a flurry of events we'll end up processing them after
-                // they've all come in.
-                var cancellationToken = _tokenSource.Token;
-                Task.Delay(TimeSpan.FromSeconds(1), cancellationToken)
-                    .ContinueWith(
-                        async _ =>
-                        {
-                            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
-
-                            ProcessBatchedChangesOnForeground(cancellationToken);
-                        },
-                        cancellationToken,
-                        TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
-                        TaskScheduler.Default).Unwrap();
-            }
-        }
-
-        private void ProcessBatchedChangesOnForeground(CancellationToken cancellationToken)
-        {
-            this.AssertIsForeground();
-
-            // If we've been asked to stop, then there's no point proceeding.
-            if (cancellationToken.IsCancellationRequested)
-            {
-                return;
-            }
+            ThisCanBeCalledOnAnyThread();
 
             // If we've been disconnected, then there's no point proceeding.
             if (_workspace == null || !IsEnabled)
-            {
                 return;
-            }
 
-            // Get a project to process.
+            // Figure out the entire set of projects to process.
+            using var _ = PooledHashSet<ProjectId>.GetInstance(out var projectsToProcess);
+
             var solution = _workspace.CurrentSolution;
-            var projectId = DequeueNextProject(solution);
-            if (projectId == null)
+            AddProjectsToProcess(workQueue, solution, projectsToProcess);
+
+            // And Process them one at a time.
+            foreach (var projectId in projectsToProcess)
             {
-                // No project to process, nothing to do.
-                return;
+                cancellationToken.ThrowIfCancellationRequested();
+                await ProcessProjectChangeAsync(solution, projectId, cancellationToken).ConfigureAwait(false);
```

Comment:
any reason this needs to be done serially other than accessing the ```_projectToInstalledPackageAndVersion``` - could we instead do it in parallel by returning the results and updating at the end?

Reply:
in general we do not do things in parallel in VS>  that's because the threadpool is both a shared resource for many components to use, and because something behaving semi-ok, might now behave very badly in parallel.  i.e. if this call ends up usign a lot of memory, then if we parallelize it we migth totally saturate teh memory subsystem causing GCs that have a heavy impact on the typing experience.

In OOP we flip things.  We're not sharing the GC with anyone else.  Our operations tend to be throughput oriented.  And having a pause for GC is far less impactful versus the speedup gains we get running in parallel.